lxd/certificates.go

```diff
@@ -500,10 +507,24 @@ func certificatesPost(d *Daemon, r *http.Request) response.Response {
 					return response.Forbidden(fmt.Errorf("No matching cluster join operation found"))
 				}
 			} else {
-				// Otherwise check if password matches trust password.
-				if util.PasswordCheck(secret, req.Password) != nil {
-					logger.Warn("Bad trust password", log.Ctx{"url": r.URL.RequestURI(), "ip": r.RemoteAddr})
-					return response.Forbidden(nil)
+				// Check if client join token supplied as password.
+				joinToken, err := shared.ClientJoinTokenDecode(req.Password)
+				if err == nil {
+					// If so then check there is a matching join operation.
+					joinOp, err := clientJoinTokenValid(d, r, project.Default, joinToken)
+					if err != nil {
+						return response.InternalError(errors.Wrapf(err, "Failed during search for join token operation"))
+					}
+
+					if joinOp == nil {
+						return response.Forbidden(fmt.Errorf("No matching client join operation found"))
+					}
```

Comment:
I'm failing to see where we're pulling the `req` from the token creation here. It looks like as soon as we found a valid token, the user can feed us whatever they want through `req` and we'll just add it to the trust store.

This would be a huge security issue. It's very important that what's described in https://discuss.linuxcontainers.org/t/lxd-token-based-remote-connection/13114/2?u=stgraber is implemented.

Reply:
> It looks like as soon as we found a valid token, the user can feed us whatever they want through `req` and we'll just add it to the trust store.

That would be possible. In order to prevent this, we could simply override `req`, can't we?

```go
req = api.CertificatesPost{
	CertificatePut: api.CertificatePut{
		Name: joinToken.ClientName,
		Type: "client",
	},
}
```

Reply:
We'd also need to add `Restricted` and `Projects` as I missed those.

Reply:
Right, we need to completely discard the user provided req and make a new one from the operation metadata.

And indeed, restricted and projects are the more critical fields here as it's they get omitted or the user can override them, then they can escalate privileges and get control of the entire system.